libs/pika/async_mpi/include/pika/async_mpi/transform_mpi.hpp

```diff
@@ -109,198 +151,259 @@ namespace pika::mpi::experimental {
                 using result_type = mpi_request_invoke_result_t<F, Ts...>;
                 using type =
-                    std::conditional_t<std::is_void<result_type>::value,
-                        Tuple<>, Tuple<result_type>>;
+                    std::conditional_t<std::is_void<result_type>::value, Tuple<>, Tuple<result_type>>;
             };
 
-            template <template <typename...> class Tuple,
-                template <typename...> class Variant>
-            using value_types =
-                pika::util::detail::unique_t<pika::util::detail::transform_t<
-                    typename pika::execution::experimental::sender_traits<
-                        Sender>::template value_types<Tuple, Variant>,
+            // -----------------------------------------------------------------
+            // get pack of unique types from combined tuple and variant
+            template <template <typename...> class Tuple, template <typename...> class Variant>
+            using value_types = pika::util::detail::unique_t<
+                pika::util::detail::transform_t<typename pika::execution::experimental::sender_traits<
+                                                    Sender>::template value_types<Tuple, Variant>,
                     invoke_result_helper>>;
 
             template <template <typename...> class Variant>
-            using error_types =
-                pika::util::detail::unique_t<pika::util::detail::prepend_t<
-                    typename pika::execution::experimental::sender_traits<
-                        Sender>::template error_types<Variant>,
-                    std::exception_ptr>>;
+            using error_types = pika::util::detail::unique_t<pika::util::detail::prepend_t<
+                typename pika::execution::experimental::sender_traits<Sender>::template error_types<Variant>,
+                std::exception_ptr>>;
 
             static constexpr bool sends_done = false;
 #endif
 
+            struct priority_set_restore
+            {
+                pika::execution::thread_priority old_priority_;
+                priority_set_restore(pika::execution::thread_priority new_p)
+                  : old_priority_{threads::detail::get_self_id_data()->get_priority()}
+                {
+                    threads::detail::get_self_id_data()->set_priority(new_p);
+                }
+
+                ~priority_set_restore()
+                {
+                    threads::detail::get_self_id_data()->set_priority(old_priority_);
+                }
+            };
+
+            // -----------------------------------------------------------------
+            // operation state for a given receiver
             template <typename Receiver>
             struct operation_state
             {
                 std::decay_t<Receiver> receiver;
                 std::decay_t<F> f;
                 stream_type stream;
-
+                pika::spinlock mutex_;
+                pika::condition_variable cond_var_;
```

Comment:
I know this is not enforced across the codebase, but let's try to keep it consistent inside a single class at least:
```suggestion
                pika::spinlock mutex;
                pika::condition_variable cond_var;
```